### ISSUE TITLE
Adopt f-strings

### ISSUE THREAD

**Issue by chris-rands** [biopython/biopython#2510]
Proposal for discussion and testing to adopt f-strings, new in Python 3.6 (see python docs and PEP 498), instead of `str.format()` and decimal style string formatting. See this and subsequent comments https://github.com/biopython/biopython/pull/2450#issuecomment-570091366

`pyupgrade --py36-plus` is one tool for automatically converting to f-strings, but after some minor testing, I think that [`flynt`](https://github.com/ikamensh/flynt) may be more comprehensive (and will only work on f-strings). If we proceed, I suggest PRs are submitted one module (or directory) at a time

**Comment by peterjc** [comment]
It still feels a bit alien right now, but for the simple ``"Hello %s, how are you?" % name`` style examples with just ``%s`` default string formatting, ``f"Hello {name}, how are you?"`` is clear and concise. I think I could get used to that pretty quickly.

Also, it finally makes the ``__format__`` methods far more useful (via the colon syntax), e.g.

```python
>>> from Bio import SeqIO
>>> record = SeqIO.read("lavender.nu", "fasta")
>>> print(f"Our record ID {record.id} in FASTA format:\n\n{record:fasta}")
```
```
Our record ID gi|5690369|gb|AF158246.1|AF158246 in FASTA format:

>gi|5690369|gb|AF158246.1|AF158246 Cricetulus griseus glucose phosphate isomerase (GPI) gene, partial intron sequence
GGCTCTTAAGTCATGTCTAGGCAGGTGTGCACAAGTTTAGGTCTTGTTCCGCCAGCATCT
GAGGAGATTGGGTTGGGAACTGGTATGCATGTCACCATCTGAGTTTAGCCCAGATCGGGA
GTCCTCTTCATCAGCATTGAGAAAGTAGAGACTGGCACAGGTCATAATGAGGTGGGTCAA
GCATGGATATGAGGCCCTGGTTCTGCCATCACCGAGACGGCAAAGGCTAGCATAGGTATG
AGACTTCCTTCCTGCCAGTGCTGAGGAACTGGGAGCCTACAGAGGTTTGAGGCCTTGGGT
CTGCCAGAAGAAAGGCCAGCTGCGGTAACATCTTTGACCGTCCCTCACGTAGTCTGCCTT
GCTGGCCAGCATGAAGAGTCACTAGTTCCTGGATCTGCTGTCTGAACTGGGAAGAGTCGG
GTTGGGTCCACCCTCCAGGCTGTTACCTGTGGAAGATTGGCTTTCTGGGGACAGGTGTGG
CAGGAGGGAGGGGGGTAGGGTGTGCCTGCCCTCCTCCAGTGCCGTTTCCTTCTCTCCTAG
TGGGTAGGTG
```

**Comment by MarkusPiotrowski** [comment]
With all this Python 2/3 changes the last three weeks we should maybe slow down a little bit and ask what is really neccesary. I think there is no problem in using f-strings if you want to, but why should we remove/convert all % and .format string formattings? Isn't it valid code anymore? Is it against PEP8 or PEP257?

**Comment by peterjc** [comment]
I'm not sure how often PEP8 gets updated with language changes, but the current version pre-dates f-strings. I wouldn't be surprised to see it updated to recommend them.

If our consensus is to go for f-strings, I would say get the conversion done ASAP as part of applying black and all the Python 2 cleanup. Cross reference my proposal to stop at that point for Biopython 1.77 (a largely spring cleaning release):

https://mailman.open-bio.org/pipermail/biopython/2020-January/016745.htmles.

**Comment by peterjc** [comment]
Interestingly in answer to Markus' question about the older string formatting options, PEP3101 is vague, suggesting at some point the older systems would be deprecated:

> Backwards Compatibility
> Backwards compatibility can be maintained by leaving the existing mechanisms in place. The new system does not collide with any of the method names of the existing string formatting techniques, so both systems can co-exist until it comes time to deprecate the older system.

Interestingly the Python 3.0 release notes said:

> PEP 3101: A New Approach To String Formatting
> A new system for built-in string formatting operations replaces the % string formatting operator. (However, the % operator is still supported; it will be deprecated in Python 3.1 and removed from the language at some later time.) Read PEP 3101 for the full scoop.

https://docs.python.org/3/whatsnew/3.0.html#pep-3101-a-new-approach-to-string-formatting

However, that deprecation didn't happen, and I've not found any similar statement in the Python 3.1 to 3.8 release notes.

https://www.python.org/dev/peps/pep-3101/

Apparently there are no current plans to deprecate the percent operator for string interpolation:

https://bugs.python.org/issue14123
https://github.com/python/cpython/commit/3bbdc8e822b483fdfb66e0422e94630af10d1b80

It looks like Python 3 will maintain all the current string formatting, but Python 4 might make a choice.

**Comment by JoaoRodrigues** [comment]
Hi all,

I agree with Markus that there has been a lot of (good!) work on moving to
Python 3, which might be hard to follow. Those of us working on some other
features will have to rebase nearly every day to take these changes
into consideration, so it might be nice to slow down or have a roadmap of
what we want to get done. Maybe we could use the Projects tab on github to
get this more organized.

As for the f-strings specifically, I would advocate to completely replace
whatever other format there is by them.

Cheers and happy new year,

João

Peter Cock <notifications@github.com> escreveu no dia quarta, 8/01/2020
à(s) 08:57:

**Comment by peterjc** [comment]
I've not used GitHub Projects before, we've not seriously ever used milestones on Biopython either (we had them on the RedMine tracker). Does anyone else think it would help?

My earlier suggestion was to finish the Python 2 cleanup, application of black to ``Bio/`` (with black on ``Tests/`` to follow later) and call that Biopython 1.77 within the next month or so:

https://mailman.open-bio.org/pipermail/biopython/2020-January/016745.htmles

Following Chris' f-string suggestion, I've tried it on another project, and overall I like it (in particular I like it a lot more than the ``.format(...)`` approach). If we were to do a complete automated switch, I'd like to do it soon as part of this spring cleaning - it is the only other large scale code cleanup on the horizon, so if we get that out of the way too, there would only be one round of updating old pull requests / feature branches.

And the Zen of Python backs picking one clear preference for string formatting.

**Comment by peterjc** [comment]
The vast majority of Biopython's string formatting uses the C-style percent operator, currently the format method is used in about 22 files, and we don't yet use the f-string style.

Pull request #2524 replaces the format method with f-strings, which could be the basis of a compromise house style?:

* C-style percent operator formatting allowed, e.g. ``"Hello %s" % name``
* String format method rejected, e.g. ``"Hello {0}".format(name)``
* f-string allowed, e.g. ``f"Hello {name}"``

**Comment by chris-rands** [comment]
If a comprise is needed, then #2524 seems like a good one to me. I am in favour of complete conversion to f-strings at some point because f-strings offer better performance, shorter syntax, generally improved readability, and consistency. However, I understand that it is pain for people to rebase ongoing PRs, and there is no rush if there is no consensus. Another large-scale, possibly harder and more disruptive, change I would also think about for future releases is adding variable annotations as suggested #2236.

**Comment by peterjc** [comment]
I'm broadly in favour of variable annotations (#2236), but as far as I know there are no automated tools for that - for now it would down to module owners to start annotating their code. i.e. This would be very gradual.

With f-strings (if we go for it as the preferred style), we have the option of doing the entire code base quickly (could even be a single monster commit, but breaking it up would be more sensible for reviewing).

**Comment by peterjc** [comment]
@mdehoon approved #2524 (replacing the string format method uses with f-strings). I propose to merge that if there are no objections over the weekend?

**Comment by MarkusPiotrowski** [comment]
Just to clarify, I really find the new f-strings cool and will for sure use them, but I don't like the compromise.  You can either argument that

- all three formatting styles are a valid Python 3, none is deprecated and none is recommended to be avoided by PEP8 (and even black will not touch it, https://github.com/psf/black/issues/363). So there is no harm in using all of them

or 

- f-strings are clearly superior to the other formatting options, use only them (Zen of Python)

I prefer the first option but can live with the second also. However, I can't see why %-formatting is preferable over using `.format`. So why should we allow the one but not the other?

The problem I see is that we may be going towards a rather arbitrary coding style which becomes hard to follow, especially for newcomers. We already try to follow (most of) PEP8 and PEP257 and we (are going) to use black. This is great. But if we add numerous additional rules this may become difficult (even if there exist flake8 plugins).

**Comment by peterjc** [comment]
Changing our small amount of format method usage is quite minor, and I find switching those to f-strings an improvement (#2424). Similarly, replacing the format method with classical %-formatting would also make the code more readable _to me_, but that doesn't seem like a good long term plan, so I'm not seriously suggesting this.

Other than historical familiarity and avoiding code churn, is there any reason to prefer %-formatting over f-strings? Joao and others are rightly concerned switching them all to f-strings could impact pre-existing branches and pull requests. I think they'll mostly need updating anyway with the Python 2/3 and black changes, so we might as well do the f-string formatting now too.

I've not found a flake8 plugin, but I if we did adopt f-strings then flynt has a check mode which could be used for continuous integration.

**Comment by peterjc** [comment]
This is still experimental, but since we might want it I started work on a flake8 plugin for string formatting style, https://github.com/peterjc/flake8-sfs - following using v0.0.2:

```
$ flake8 --select SFS --statistics BioSQL Bio 
...
1518  SFS101 String literal formatting using percent operator.
172   SFS201 String literal formatting using format method.
```

It didn't find any f-strings (although there are a couple in our docstring examples now), but the statistics give an indication of the relative usage right now.

**Comment by chris-rands** [comment]
I would be in favour of adopting f-strings and using Peter's nice looking plugin to enforce it, unless these choices really do put people off from contributing (I know if can be frustrating dealing with travis failures due to minor `flake8` issues)

`black` doesn't do f-string conversion because it's out of scope, but for `black`s own source code in their github repo, they actually have chosen to use f-strings only I think.

**Comment by svalqui** [comment]
I'll work on Bio.Uniprot, as I'm reviewing GOA.py

**Comment by JoaoRodrigues** [comment]
+1 for complete f-string conversion (and using an automated tool if there
is one!).

Happy to review changes to Bio.PDB.

Sergio <notifications@github.com> escreveu no dia terça, 14/01/2020 à(s)
14:03:

**Comment by peterjc** [comment]
The experiment with flynt on #2524 showed that with its default setting it does not convert everything - in particular it avoids making overly long f-string lines. You can tell it to ignore that, and then manually break the long lines (or we could ignore the long lines in the short term - we don't yet enforce a line length via flake8). See fcf7ee6b970915488b1ad756b2213c74066ad4ab for a concrete example of this.

Also, on another project, where I ran flynt for percent conversion (not that we've agreed about doing this), I've found flynt misses lots of corner cases.

i.e. flynt is not a comprehensive solution, although it will take case of a lot of the manual work if we were to want to do a large scale move to f-strings.

**Comment by morrme** [comment]
@peterjc what is the status of this issue?

**Comment by peterjc** [comment]
Current counts on the master branch,

```
$ flake8 --isolated --select SFS --statistics BioSQL Bio
...
1540  SFS101 String literal formatting using percent operator.
30    SFS201 String literal formatting using format method.
164   SFS301 String literal formatting using f-string.
```

Looks like some format method usage snuck back in?

**Comment by peterjc** [comment]
For reference accepting all of the pyupgrade 2.25.0  suggestions (many of which I currently feel are less readable as it likes the ``.format(...)`` method), converts just over a quarter of the old style % strings:

```
$ pyupgrade --py37-plus `find Bio BioSQL -name "*.py"`
$ black Bio BioSQL
$ flake8 --isolated --select SFS --statistics BioSQL Bio
...
1195  SFS101 String literal formatting using percent operator.
99    SFS201 String literal formatting using format method.
437   SFS301 String literal formatting using f-string
```

Alternatively, using flynt 0.69 defaults:

```
$ flynt Bio BioSQL
$ black Bio BioSQL
$ flake8 --isolated --select SFS --statistics BioSQL Bio
..
505   SFS101 String literal formatting using percent operator.
27    SFS201 String literal formatting using format method.
1202  SFS301 String literal formatting using f-string
```

We can convert even more if we let flynt make long lines, and then perhaps se black's experimental ``--experimental-string-processing`` option.

**Comment by peterjc** [comment]
I wonder if there are any off the shelf ways to count string formatting within doctests (and auto-update them)?
